src/fairseq2/recipes/hf/datasets.py

```diff
@@ -0,0 +1,188 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+from typing import Any, Dict, Optional, Union, final
+
+from datasets import Dataset  # type: ignore
+
+from fairseq2.assets.card import AssetCard
+from fairseq2.data.data_pipeline import Collater, create_bucket_sizes, read_sequence
+from fairseq2.datasets.data_reader import DataPipelineReader
+from fairseq2.datasets.loader import DatasetLoader
+from fairseq2.gang import Gang
+
+HFBatch = Dict[str, Any]
+
+
+class HFDataset:
+    def __init__(self, name: str):
+        self.name = name
+    
+    def __repr__(self) -> str:
+        return self.name
+
+    def create_reader(
+        self,
+        dataset: Dataset,
+        gang: Gang,
+        bucket_size: Optional[int] = None,
+        max_num_elements: Optional[int] = None,
+        max_seq_len: Optional[int] = None,
+        min_seq_len: Optional[int] = None,
+        seq_len_col: Optional[str] = None,
+        num_accumulate: int = 1,
+        num_prefetch: int = 1,
+        seed: int = 2,
+        **extra: Any,
+    ) -> DataPipelineReader[HFBatch]:
+        """
+        Convert HF Dataset into a fairseq2 PipelineReader.
+        The HF-specific processing logic is assumed to be
+        done before the conversion.
+
+        **Batching**: The batching can be done externally or using
+        bucketing strategies in fairseq2. If both ``bucket_size``
+        and ``max_num_elements`` are None, then we assume the
+        ``dataset`` is already batched and do nothing. Otherwise:
+
+        - If ``bucket_size`` is not None, generate batches of fixed size
```

Comment:
Wondering whether it would be cleaner if we just had two separate functions. One using `bucket_by_length`, the other `bucket`. `bucket_size` and `max_num_elements` are mutually exclusive, so they feel a bit out of place when taken as parameters in a single function.

Reply:
ok, I create a `Batcher` interface that act as a "visitor" to be applied on the data pipeline. Hope this modularization is enough.